pkg/kube/krt/bench_test.go

```diff
@@ -0,0 +1,244 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package krt_test
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"testing"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	klabels "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+
+	"istio.io/istio/pkg/config/labels"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/controllers"
+	"istio.io/istio/pkg/kube/kclient"
+	"istio.io/istio/pkg/kube/kclient/clienttest"
+	"istio.io/istio/pkg/kube/krt"
+	"istio.io/istio/pkg/slices"
+	"istio.io/istio/pkg/test"
+)
+
+type Workload struct {
+	krt.Named
+	ServiceNames []string
+	IP           string
+}
+
+func NewModern(c kube.Client, events chan string, stop <-chan struct{}) {
+	Pods := krt.NewInformer[*v1.Pod](c)
+	Services := krt.NewInformer[*v1.Service](c)
+	Workloads := krt.NewCollection(Pods, func(ctx krt.HandlerContext, p *v1.Pod) *Workload {
+		if p.Status.PodIP == "" {
+			return nil
+		}
+		services := krt.Fetch(ctx, Services, krt.FilterNamespace(p.Namespace), krt.FilterSelectsNonEmpty(p.GetLabels()))
+		return &Workload{
+			Named:        krt.NewNamed(p),
+			IP:           p.Status.PodIP,
+			ServiceNames: slices.Map(services, func(e *v1.Service) string { return e.Name }),
+		}
```

Comment:
This is a good example. my doubt can this `Workloads` be notified when service update? Like a service labels updated, it may not select the pod now. 

How can it be achieved

Reply:
That is the whole point of the library :slightly_smiling_face: . It will automatically handle this for you.

How it works, under the hood, is it builds up a mapping of the dependencies. For example `Pod/foo -> Service{filter={selects app=foo}}`. When a service changes, it will notify the Collection, which will run through the dependencies (`onSecondaryDependencyEvent`) and find all the pods that depended on it (from the old or new object). Any that were impacted, it will rerun the transformation function and recompute the Workload. If the Workload is the same, it will stop; else it will call handlers.

So overall, the point of the library is that users can make a declarative function (Pod + Dependencies -> Workload), and the library will handle all the event and state management. So this answers https://github.com/istio/istio/pull/47891#discussion_r1398790195 as well I hope; I partially wrote this benchmark to show side by side the old way vs krt way of writing the same controller to help show the value it provides.

Actually the old way is more complex than this example shows, because I have (multiple!) bugs in the old controller -- which just goes to show how easy it is to make mistakes.